net: assert CNode::m_inbound_onion is inbound in ctor, add getter, unit tests

jonatack opened the pull request:
The goal of this PR is to be able to depend on `m_inbound_onion` in AttemptToEvictConnection in #20197:

- asserts `CNode::m_inbound_onion` is inbound in the CNode ctor to have a validity check at the class boundary
- fixes a unit test and a fuzz utility that were passing invalid inbound onion values to the CNode ctor
- drops an unneeded check in `CNode::ConnectedThroughNetwork()` for its inbound status
- adds a public getter `IsInboundOnion()` that also allows unit testing it
- adds unit test coverage

jonatack committed:
net: assert CNode::m_inbound_onion is inbound in ctor

and drop an unneeded check in CNode::ConnectedThroughNetwork()

jonatack committed:
net: add CNode::IsInboundOnion() public getter and unit tests

hebasto reviewed (approved):
ACK 39286b3c6f2bae7d3ea77185689469bb22572189, tested on Linux Mint 20 (x86_64).

practicalswift commented:
Concept ACK

Thanks for improving testing!

jonatack commented:
Rebased.

theStack commented on src/net.h:
```diff
@@ -1095,7 +1095,7 @@ class CNode
 
-    //! Whether this peer connected via our Tor onion service.
+    //! Whether this peer is an inbound onion, e.g. connected via our Tor onion service.
     const bool m_inbound_onion{false};
```
yocto-nit (not directly related to this PR though, feel free to ignore): not sure if we have any guidelines (or dominant opinions) on this subject, but isn't it overkill to have two initializations for a class member: once via in-class initialization and once via initializer list in the ctor? (And yes, I'm aware that initializer list have higher priority in C++). As reader of the class interface I'd not assume an in-class initialization is overruled in the ctor. Also, without the in-class initialization the compiler has the chance to throw an error if the const member was forgot to be initializer.

jonatack replied:
Hm, good eye.

-> In this case, ISTM we could remove the default `false` argument in the constructor declaration

-> As to your last point, it suggests the default member `m_inbound_onion` initializer not set a value, since we are doing it dynamically in the ctor. It's true that I see a build error in this case, but not without the change in `net.h:1099`
```diff
/src/net.cpp
-    nMyStartingHeight(nMyStartingHeightIn),
-    m_inbound_onion(inbound_onion && conn_type_in == ConnectionType::INBOUND)
+    nMyStartingHeight(nMyStartingHeightIn)

/src/net.h
     //! Whether this peer is an inbound onion, e.g. connected via our Tor onion service.
-    const bool m_inbound_onion{false};
+    const bool m_inbound_onion;
```
```
net.cpp:2947:1: error: uninitialized const member in ‘const bool’ [-fpermissive]
 2947 | CNode::CNode(NodeId idIn, ServiceFlags nLocalServicesIn, int nMyStartingHeightIn, SOCKET hSocketIn, const CAddress& addrIn, uint64_t nKeyedNetGroupIn, uint64_t nLocalHostNonceIn, const CAddress& addrBindIn, const std::string& addrNameIn, ConnectionType conn_type_in, bool inbound_onion)
      | ^~~~~
In file included from net.cpp:10:
./net.h:1099:16: note: ‘const bool CNode::m_inbound_onion’ should be initialized
 1099 |     const bool m_inbound_onion;
      |                ^~~~~~~~~~~~~~~
```
I don't mind pulling in a commit from your branch if you'd like to adjust this.

-----

More generally, from what I can gather, the ideas behind in-class default member initializers added in C++11 were (a) to allow a non-static data member to be initialized where it is declared, e.g. in its class. A constructor can then use the initializer when run-time initialization is needed, which can then (b) tidy up the code up by avoiding doing it in the member initializer list of the constructor. This provides extra benefits in classes with multiple constructors.

I guess one can see the default member initializer as the general default, particularly useful for constant initializers, that can be overridden by any specific ones in member initializer lists in the constructors.

Some resources (send me any good ones you might suggest; my understanding on this is evolving):
- https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines.html#Rc-in-class-initializer
- https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines.html#c49-prefer-initialization-to-assignment-in-constructors
- https://en.cppreference.com/w/cpp/language/data_members#member-initialization
- https://www.stroustrup.com/C++11FAQ.html#member-init

jonatack replied:
(Or you could open a separate PR, for as you mentioned, it's not directly related to this change. Either way SGTM.)

jonatack replied:
> not sure if we have any guidelines (or dominant opinions) on this subject, but isn't it overkill to have two initializations for a class member: once via in-class initialization and once via initializer list in the ctor? (And yes, I'm aware that initializer lists have higher priority in C++). As reader of the class interface I'd not assume an in-class initialization is overruled in the ctor. Also, without the in-class initialization the compiler has the chance to throw an error if the const member was forgot to be initialized in the ctor.

@theStack: done in #21167

theStack reviewed (approved):
ACK 0a30f5d9b44abaed2f08f33ad2bcaa20ab5cbdba 🏞️

laanwj commented on src/net.cpp:
```diff
@@ -2957,7 +2957,7 @@ CNode::CNode(NodeId idIn, ServiceFlags nLocalServicesIn, int nMyStartingHeightIn
     nLocalServices(nLocalServicesIn),
     nMyStartingHeight(nMyStartingHeightIn),
-    m_inbound_onion(inbound_onion)
+    m_inbound_onion(inbound_onion && conn_type_in == ConnectionType::INBOUND)
```
I might be misunderstanding but in what case would the caller pass `inbound_onion` parameter as true but lie about this, as it is an outbound connection?
Wouldn't this be better as an assertion?

MarcoFalke replied:
Or an `Assume`, if `Assert` is too strong and risky

jonatack replied:
Good idea, `inbound_onion` should only be true when a new CNode is instantiated from `CConnman::AcceptConnection()`, so this should only fail if a bug is added. Made it an assert.

jonatack replied:
> in what case would the caller pass `inbound_onion` parameter as true but lie about this

So, adding the assertion found one! Our net_tests had a case that lied; updating it.

jonatack replied:
Hm, a fuzzer needs to be updated, too.

jonatack commented:
Updated (and rebased for the CI) per `git range-diff cfbfd38 0a30f5d dd01a23`

jonatack committed:
test, fuzz: fix constructing CNode with invalid inbound_onion

as CNode ctor should only be passed inbound_onion = true
when the connection is inbound

jonatack commented:
Rebased/simplified after merge of #20686.

sipa commented:
utACK 86c495223f048e5ca2cf0d8730af7db3b76f7aba

LarryRuane reviewed (commented):
ACK 86c495223f048e5ca2cf0d8730af7db3b76f7aba
Compiled, ran unit tests.

vasild commented on src/test/fuzz/util.h:
```diff
@@ -298,7 +298,7 @@ CNode ConsumeNode(FuzzedDataProvider& fuzzed_data_provider) noexcept
     const std::string addr_name = fuzzed_data_provider.ConsumeRandomLengthString(64);
     const ConnectionType conn_type = fuzzed_data_provider.PickValueInArray({ConnectionType::INBOUND, ConnectionType::OUTBOUND_FULL_RELAY, ConnectionType::MANUAL, ConnectionType::FEELER, ConnectionType::BLOCK_RELAY, ConnectionType::ADDR_FETCH});
-    const bool inbound_onion = fuzzed_data_provider.ConsumeBool();
+    const bool inbound_onion{conn_type == ConnectionType::INBOUND ? fuzzed_data_provider.ConsumeBool() : false};
```
nit: same as `ConnectionType::INBOUND && fuzzed_data_provider.ConsumeBool()`

jonatack replied:
indeed, will update if need to retouch or rebase

MarcoFalke replied:
difference is that it needs to store one additional redundant byte for every non-inbound fuzz input file, so I think this can be kept as is

vasild commented on src/net.h:
```diff
@@ -1234,6 +1234,9 @@ class CNode
     std::string ConnectionTypeAsString() const;
+
+    /** Whether this peer is an inbound onion, e.g. connected via our Tor onion service. */
+    bool IsInboundOnion() const { return m_inbound_onion; }
```
Given that this method is added only to be used in tests, could the tests use
`pnode->IsInboundConn() && pnode->ConnectedThroughNetwork() == NET_ONION` which is equivalent to the newly added `pnode->IsInboundOnion()`?

jonatack replied:
The goal is to test `m_inbound_onion` to be able to depend on it in #20197.

jonatack replied:
(the unit tests already assert on `IsInboundConn()`, so this would shift the coverage from `m_inbound_onion` directly, to indirectly via `ConnectedThroughNetwork() == NET_ONION`, which would avoid the getter but seems less robust?)

vasild replied:
Yeah, it will test it indirectly, less robust. Feel free to ditch this conversation :)

vasild reviewed (approved):
ACK 86c495223f048e5ca2cf0d8730af7db3b76f7aba

Two minor comments below, feel free to ignore.

MarcoFalke commented on src/net.h:
```diff
@@ -1234,6 +1234,9 @@ class CNode
 
     std::string ConnectionTypeAsString() const;
+
+    /** Whether this peer is an inbound onion, e.g. connected via our Tor onion service. */
```
nit question: should this say "i.e."? I mean are there any other ways to establish an inbound onion other than via the tor onion service?

jonatack replied:
I don't mind updating this and taking @vasild's suggestion at the same time.

MarcoFalke replied:
Can do as a follow-up. The merge script is already running and this should land in master in the next couple of minutes.

jonatack replied:
Oh ok, will do in #20210.

jonatack replied:
> nit question: should this say "i.e."? I mean are there any other ways to establish an inbound onion other than via the tor onion service?

Done in #21167

maflcko reviewed (approved):
review ACK 86c495223f048e5ca2cf0d8730af7db3b76f7aba 🐍

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

review ACK 86c495223f048e5ca2cf0d8730af7db3b76f7aba 🐍
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUjcBAwAyH/ZAIoBmnAriDbYf2IdudHCi55YS3Gb2Viv4ujVLCCt+dTDj6OOg6RR
fsTKJe3f0g3eHdss1KkMxTGQm1WP6MDl7Iqpkuuar7eH4F7WslFSamnnRStDZXcG
vUQvrGSKNakc20PT0wT8bP7UxA2+z8/gVS9V40+lKma8VmwZgYE0EcKztdfUiPL5
0v+whP/aUPxjEleeG4aec0FkOEwELAdoC6gTtn2+hBBv2x+Kj6KmeheusTNBpP7P
jYbMqz5K+qEd10yFc+4yH7gdVyKmAuNunToURLz2SZcQmla2iMinRIXboLEhRcvm
pS4YuY62sQ+3bOxxTi3Td7vRMKFmDJfy92vkXgDcQvvpHw72M+k+3ieuzYNvJFUq
maOT4trYzF79M5F4Pd18ywIkgz8v50uddPVEVhl4LGeZyAN47WOH0vIoSZBhmuNN
IdGom1GuBOMsztoXEZDLYEPr9X9HDB8CgGW5t02g4cf/EoqbtZHzIM2VOXDkiVYe
Y5IbMXDV
=O80S
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `683c8f1c6754ede192ead999e54c0440e14b98b058b5f69d6e0b9817be2c17bb  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e892940108683c8f1c6754ede192ead999e54c0440e14b98b058b5f69d6e0b9817be2c17bbf0104a570b70523d498ca9b44a4b05737e0308fff0101ab89e6a6bc08df3f1e76ba56ca57eaf08f1045ff033f2f00872c617c06a7e3db30083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6dfff0107a3a108f4bb2afddecd43016320fc65208f1045ff033f2f008f4b779579312ef630083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff010aa023b5dd2439de9a8de63aaeef63d5d08f120c72a40a3cfd1bc941cbc0eaa434a7b7902e9881733f7b84560862a6e2e77055008f1045ff033f1f008fb2f7339709355b60083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267f010f6a8254f48e695f79c7a3958d4d94ced08f1045ff033f1f0083bd4e51e5f2943e50083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6d)

</details>

jnewbery commented:
Is there any benefit to adding getter boilerplate code to the header class instead of just making this const member public?

jonatack commented:
@jnewbery Unless additional semantics will be needed for future changes or refactoring, [no](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines.html#c131-avoid-trivial-getters-and-setters), probably not. I was being prudent.

jnewbery commented:
oooh thanks for referencing the cpp core guidelines! I'd say that any getter of a const value is trivial :)

jonatack commented:
Your comment is why I found it ;)

jonatack commented:
> oooh thanks for referencing the cpp core guidelines! I'd say that any getter of a const value is trivial :)

@jnewbery thanks! done in #21167
